go/internal/cbor/deterministic.go

```diff
@@ -0,0 +1,121 @@
+package cbor
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+)
+
+// Deterministic function loops through the given byte array containing CBOR sequence
+// and checks that they follow the deterministic principles described here:
+// https://www.rfc-editor.org/rfc/rfc8949.html#name-deterministically-encoded-c.
+func Deterministic(input []byte) error {
+	index := 0
+
+	for index < len(input) {
+		// TODO(sonkkeli): Remove debug print.
+		fmt.Printf("index=%v\n", index)
+
+		length, err := deterministicRec(input[index:])
+		if err != nil {
+			return err
+		}
+		index += length
+	}
+	return nil
+}
+
+// deterministicRec is a recursively called helper function to check the deterministicy of a bytes array.
+// TODO(sonkkeli): Recursively called parts will be the maps and arrays, which contain other types, meaning
+// unsigned integers, byte strings, text strings.
+func deterministicRec(input []byte) (int, error) {
+	cborItemLengthInBytes := 0
+
+	b := input[0]
+
+	switch getMajorType(b) {
+	case TypePosInt:
+		length, _, err := unsignedIntegerDeterministic(input)
+		if err != nil {
+			return 0, err
+		}
+		cborItemLengthInBytes = length
+
+		break
+
+	case TypeBytes, TypeText:
+		// TODO(sonkkeli):
+		// length, err := textOrBytesDeterministic(input[index:])
+		break
+
+	case TypeArray:
+		// TODO(sonkkeli):
+		// length, err := arrayDeterministic(input[index:])
+		break
+
+	case TypeMap:
+		// TODO(sonkkeli):
+		// length, err := mapDeterministic(input[index:])
+		break
+
+	default:
+		return 0, errors.New("Missing implementation for major type.")
+	}
+
+	// TODO(sonkkeli): Remove debug print.
+	fmt.Printf("cborItemLengthInBytes=%v\n", cborItemLengthInBytes)
+
+	return cborItemLengthInBytes + 1, nil
+}
+
+// unsignedIntegerDeterministic calculates the value of the unsigned integer to ensure that
+// the right amount of bytes is used in the CBOR encoding and returns the length of the unsigned integer.
```

Comment:
```suggestion
// the right amount of bytes is used in the CBOR encoding.
```
(The return values are explained in the next sentence)